here-naksha-app-service/src/test/java/com/here/naksha/app/service/ReadFeaturesWithApiFaultsHttpStorageTest.java

```diff
@@ -0,0 +1,149 @@
+    setupSpaceAndRelatedResources(nakshaClient, "ReadFeatures/WithApiFaultsHttpStorage/setup");
+  }
+
+  @Test
+  void tc01_testTimeout() throws Exception {
+    // Validate behavior on response exceeding timeout
+    String streamId = UUID.randomUUID().toString();
+
+    UrlPattern endpointPath = urlPathEqualTo(ENDPOINT + "/1");
+    stubFor(get(endpointPath).willReturn(okJson(CORRECT_ENDPOINT_RESPONSE).withFixedDelay(EXCEED_TIMEOUT_MILLIS)));
+
+    // When: Get Features request is submitted to NakshaHub Space Storage instance
+    HttpResponse<String> response = getNakshaClient().get("hub/spaces/" + SPACE_ID + "/features/1", streamId);
+
+    // Then: Perform assertions
+    ResponseAssertions.assertThat(response)
+            .hasStatus(500)
+            .hasStreamIdHeader(streamId);
+
+    assertTrue(response.body().contains("request timed out"));
+
+    // Then: Verify request reached endpoint once
+    verify(1, getRequestedFor(endpointPath));
+  }
+
+  @Test
+  void tc02_testTimeoutWithChunkedResponse() throws Exception {
+    // Validate behavior on response split to chunks. First chunk arrives on time but the whole response exceeds timeout.
+    String streamId = UUID.randomUUID().toString();
+    UrlPattern endpointPath = urlPathEqualTo(ENDPOINT + "/2");
+    stubFor(get(endpointPath).willReturn(okJson(CORRECT_ENDPOINT_RESPONSE).withChunkedDribbleDelay(5, EXCEED_TIMEOUT_MILLIS)));
+
+    // When: Get Features request is submitted to NakshaHub Space Storage instance
+    HttpResponse<String> response = getNakshaClient().get("hub/spaces/" + SPACE_ID + "/features/2", streamId);
+
+    // Then: Perform assertions
+    ResponseAssertions.assertThat(response)
+            .hasStatus(500)
+            .hasStreamIdHeader(streamId);
+
+    assertTrue(response.body().contains("request timed out"));
+
+    // Then: Verify request reached endpoint once
+    verify(1, getRequestedFor(endpointPath));
+  }
+
+  @Test
+  void tc03_testMalformedResponseChunk() throws Exception {
+    // Validate behavior on scenario when server send an OK status header, then garbage, then close the connection.
+    String streamId = UUID.randomUUID().toString();
+
+    UrlPattern endpointPath = urlPathEqualTo(ENDPOINT + "/3");
+    stubFor(get(endpointPath).willReturn(aResponse().withFault(Fault.MALFORMED_RESPONSE_CHUNK)));
+
+    // When: Get Features request is submitted to NakshaHub Space Storage instance
+    HttpResponse<String> response = getNakshaClient().get("hub/spaces/" + SPACE_ID + "/features/3", streamId);
+
+    // Then: Perform assertions
+    ResponseAssertions.assertThat(response)
+            .hasStatus(500)
+            .hasStreamIdHeader(streamId);
+
+    assertTrue(response.body().contains("java.io.IOException: protocol error"));
+
+    // Then: Verify request reached endpoint once
+    verify(1, getRequestedFor(endpointPath));
+  }
+
+  @Test
+  void tc04_testConnectionReset() throws Exception {
```

Comment:
We have sufficient coverage for validating error 500 here. Let us also test atleast one scenario where Naksha service returns some error other than 500 (depending on failure from FFW Adapter - any of 401/403/429/501)

Reply:
Done

Reply:
Done